### ISSUE TITLE
blacklisted_name should not tbe triggered in tests

### ISSUE THREAD

**Issue by tamaroning** [rust-lang/rust-clippy#8986]
### Summary

blacklisted_name should toralate use of such name as `foo`. 
As far as I see the source code, it tries to allow `foo` in tests, but it doesn't.

Similar to https://github.com/rust-lang/rust-clippy/issues/8758

### Lint Name

blacklisted_name

### Reproducer

I tried this code:

```rust
#![warn(clippy::blacklisted_name)]
fn main() {}

#[cfg(test)]
mod abc {
    fn _abc() {
        let foo = ();
        dbg!(foo);
    }
}

#[test]
fn bcd() {
    let foo = ();
    dbg!(foo);
}

fn _test() {
    let foo = ();
    dbg!(foo);
}

```

I saw this happen:

```
warning: use of a blacklisted/placeholder name `foo`
  --> src\main.rs:19:9
   |
19 |     let foo = ();
   |         ^^^
   |
note: the lint level is defined here
  --> src\main.rs:1:9
   |
1  | #![warn(clippy::blacklisted_name)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blacklisted_name

warning: use of a blacklisted/placeholder name `foo`
 --> src\main.rs:7:13
  |
7 |         let foo = ();
  |             ^^^
  |
note: the lint level is defined here
 --> src\main.rs:1:9
  |
1 | #![warn(clippy::blacklisted_name)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blacklisted_name

warning: use of a blacklisted/placeholder name `foo`
  --> src\main.rs:14:9
   |
14 |     let foo = ();
   |         ^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blacklisted_name

warning: use of a blacklisted/placeholder name `foo`
  --> src\main.rs:19:9
   |
19 |     let foo = ();
   |         ^^^
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blacklisted_name
```

I expected to see this happen:
No warning

### Version

```text
rustc 1.61.0-nightly (458262b13 2022-03-09)
binary: rustc
commit-hash: 458262b1315e0de7be940fe95e111bb045e4a2a4
commit-date: 2022-03-09
host: x86_64-pc-windows-msvc
release: 1.61.0-nightly
LLVM version: 14.0.0
```

### Additional Labels

@rustbot label +E-easy

**Comment by kyoto7250** [comment]
@tamaroning 

Hello.


I tried your code, but the following two cases cannot be reproduced. ([rust playrground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=63161d42dca77cd2c51a5f5f37946a8b))

```rust
#![warn(clippy::blacklisted_name)]
fn main() {}

#[cfg(test)]
mod abc {
    fn _abc() {
        let foo = ();
        dbg!(foo);
    }
}

#[test]
fn bcd() {
    let foo = ();
    dbg!(foo);
}
```


last case, I think we should not lint.
They should be ignored in test cases only.

`#[cfg(test)]` and `#[test]` are codes that are ignored except when the test is run,   but I don't think the name `fn test` has such a restriction.

**Comment by tamaroning** [comment]
@kyoto7250 
> I tried your code, but the following two cases cannot be reproduced.

Playground seems not to pass --tests flag to the compiler so I ran clippy in local enviroment.

> ..., but I don't think the name `fn test` has such a restriction.

Yes. I reconsider this and agree with you. But the lint has been fixed in PR https://github.com/rust-lang/rust-clippy/pull/7379 in order to trigger on `fn test`.

**Comment by kyoto7250** [comment]
@tamaroning 

I can reproduce now, thanks!

 It doesn't seem urgent, so I hope someone interested will do it.

**Comment by yerke** [comment]
I think the lint is working as intended right now.

For the following input:
```rust
#![warn(clippy::blacklisted_name)]
fn main() {}

#[cfg(test)]
mod abc {
    fn _abc() {
        let foo = ();
        dbg!(foo);
    }
}

#[test]
fn bcd() {
    let foo = ();
    dbg!(foo);
}

fn _test() {
    let foo = ();
    dbg!(foo);
}

mod tests {
    fn _mf() {
        let foo = ();
        dbg!(foo);
    }
}
```

I get only 1 warning:
```rust
warning: use of a blacklisted/placeholder name `foo`
  --> src/main.rs:19:9
   |
19 |     let foo = ();
   |         ^^^
   |
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![warn(clippy::blacklisted_name)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#blacklisted_name

warning: `clippy-example-1` (bin "clippy-example-1") generated 1 warning
```

Which is expected, since the function `_test` is a standalone function and is not part of any tests.
`mod tests` is ignored, since it contains `tests` in it.

The `is_test_module_or_function` function this lint relies on seems to work according to its description:
https://github.com/rust-lang/rust-clippy/blob/580f642cfac7d7035c563a060265b3dbef092ada/clippy_utils/src/lib.rs#L2220-L2228

Does my reasoning make sense or not?
